blockchain/validate.go

```diff
@@ -1350,51 +1447,333 @@ func (b *BlockChain) checkDupTxs(txSet []*dcrutil.Tx, view *UtxoViewpoint) error
 // NOTE: The caller MUST have already determined that the provided transaction
 // is a ticket purchase.
 func checkTicketPurchaseInputs(msgTx *wire.MsgTx, view *UtxoViewpoint) error {
-	sstxInAmts := make([]int64, len(msgTx.TxIn))
-	for idx, txIn := range msgTx.TxIn {
-		// Ensure the input is available.
+	// Assert there are two outputs for each input to the ticket as well as the
+	// additional voting rights output.
+	if (len(msgTx.TxIn)*2 + 1) != len(msgTx.TxOut) {
+		panicf("attempt to check ticket purchase inputs on tx %s which does "+
+			"not appear to be a ticket purchase (%d inputs, %d outputs)",
+			msgTx.TxHash(), len(msgTx.TxIn), len(msgTx.TxOut))
+	}
+
+	for txInIdx := 0; txInIdx < len(msgTx.TxIn); txInIdx++ {
```

Comment:
The range statement in Go is rather wonky.  It has non-obvious semantics and often leads to issues, because of the way it behaves.  For example, I'm sure you'll notice in some places that copies of things are made with comments about preventing issues that are a direct result of the terrible range semantics.

Ultimately, I would like to change the arrays of most things from being arrays of pointers to elements to arrays of elements themselves to significantly improve cache locality.  However, the range statement copies the entire element in that case, and that is, in fact, the original primary driving reason why most of the existing data structures ended up being `[]*Foo` instead of `[]Foo` so they could be used with the range statement and behave as one would expect.  I personally don't think the use of range justifies significantly slower code.